asim: final port over from mma prototype for asim

wenyihu6 opened the pull request:
**asim: pass false to ReplicaChangeDelayFn for AllocationTransferLeaseOp**

Previously, we incorrectly passed true to ReplicaChangeDelayFn for
AllocationTransferLeaseOp, causing it to treat the operation like a replica
addition and extend the delay based on range size and snapshot rate. Since lease
transfers are expected to be much faster, this was unintended. This commit fixes
the issue by applying only the base delay (ReplicaChangeBaseDelay) to
AllocationTransferLeaseOp, which defaults to 100ms.

Epic: none
Release note: none

---

**asim: correctly update queue.next in pushReplicateChange**

Previously, we did not assign the next field of the queues correctly because we
passed the base queue by value into pushReplicateChange, which caused updates to
the next field to affect only the copy, not the original queue. This commit
fixes the issue by returning the correct next value and updating the queue
directly.

Epic: none
Release note: none

---

**asim: pass simulator.Replica instead to pushReplicateChange**

Previously, we passed state.Range to pushReplicateChange. This commit changes it
to pass the simulator.Replica directly, as future commits will require access to
additional fields such as Desc and RangeUsageInfo.

Epic: none
Release note: none

---

**asim: pass state changer directly to pushReplicateChange**

Previously, we passed base queue to pushReplicateChange, but only its state
changer was being used. This commit updates pushReplicateChange to only take in
the state changer.

Epic: none
Release note: none

---

**asim: integrate allocatorsync with lease and replicate queue properly**

This commit integrates lease and replicate queue changes with the mma allocator
properly by calling allocatorSync.PreApply* and allocatorSync.PostApply,
completing the allocator integration. Some data-driven test outputs changed
because allocatorSync.PostApply now updates store pool after operations (such as
UpdateLocalStoresAfterLeaseTransfer), which was missing previously in asim. It’s
unclear if this was an oversight or previously seemed as unnecessary. More
reading is needed.

Epic: none
Release note: none

---

**asim: pass queue names for logging**

This commit passes the name of the queue making changes to pushReplicateChange
to improve logging.

Epic: none
Release note: none

---

**asim: remove setting from store**

This commit removes the setting field from store struct as it is not used
anywhere.

Epic: none
Release note: none

---

**asim: add rebalance_objective and delay for rebalance_mode**

This commit adds rebalance_objective as a setting and allows changes to
rebalance_mode to be scheduled as a delayed simulation event. This enables
testing transitions between mma enabled and disabled mode.

Epic: none
Release note: none

---

**asim: check for LBRebalancingMultiMetric in ShouldRebalanceStore**

This commit updates the old store rebalancer's ShouldRebalanceStore to check the
LBRebalancingMultiMetric cluster setting. asim still invokes this method when
checking changes from the old store rebalancer.

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/149904)
<!-- Reviewable:end -->

wenyihu6 committed:
asim: use roachpb.ReplicationTarget for AllocationTransferLeaseOp

Previously, AllocationTransferLeaseOp used store ID to represent the source and
target. This commit updates it to use roachpb.ReplicationTarget, which includes
both store id and node id. We made this change because future commits will pass
fields from AllocationTransferLeaseOp directly into NonMMAPreTransferLease,
which requires ReplicationTarget. Note that production behavior remains
unchanged, as only the store ID is accessed for them.

Epic: none
Release note: none

wenyihu6 committed:
asim: remove setting from store

This commit removes the setting field from store struct as it is not used
anywhere.

Epic: none
Release note: none

wenyihu6 committed:
asim: integrate allocatorsync with lease and replicate queue properly

This commit integrates lease and replicate queue changes with the mma allocator
properly by calling allocatorSync.PreApply* and allocatorSync.PostApply,
completing the allocator integration. Some data-driven test outputs changed
because allocatorSync.PostApply now updates store pool after operations (such as
UpdateLocalStoresAfterLeaseTransfer), which was missing previously in asim. It’s
unclear if this was an oversight or previously seemed as unnecessary. More
reading is needed.

Epic: none
Release note: none

wenyihu6 committed:
asim: add rebalance_objective and delay for rebalance_mode

This commit adds rebalance_objective as a setting and allows changes to
rebalance_mode to be scheduled as a delayed simulation event. This enables
testing transitions between mma enabled and disabled mode.

Epic: none
Release note: none

wenyihu6 committed:
asim: check for LBRebalancingMultiMetric in ShouldRebalanceStore

This commit updates the old store rebalancer's ShouldRebalanceStore to check the
LBRebalancingMultiMetric cluster setting. asim still invokes this method when
checking changes from the old store rebalancer.

Epic: none
Release note: none

wenyihu6 committed:
asim: pass false to ReplicaChangeDelayFn for AllocationTransferLeaseOp

Previously, we incorrectly passed true to ReplicaChangeDelayFn for
AllocationTransferLeaseOp, causing it to treat the operation like a replica
addition and extend the delay based on range size and snapshot rate. Since lease
transfers are expected to be much faster, this was unintended. This commit fixes
the issue by applying only the base delay (ReplicaChangeBaseDelay) to
AllocationTransferLeaseOp, which defaults to 100ms.

Epic: none
Release note: none

wenyihu6 committed:
asim: correctly update queue.next in pushReplicateChange

Previously, we did not assign the next field of the queues correctly because we
passed the base queue by value into pushReplicateChange, which caused updates to
the next field to affect only the copy, not the original queue. This commit
fixes the issue by returning the correct next value and updating the queue
directly.

Epic: none
Release note: none

wenyihu6 committed:
asim: pass simulator.Replica instead to pushReplicateChange

Previously, we passed state.Range to pushReplicateChange. This commit changes it
to pass the simulator.Replica directly, as future commits will require access to
additional fields such as Desc and RangeUsageInfo.

Epic: none
Release note: none

wenyihu6 committed:
asim: pass state changer directly to pushReplicateChange

Previously, we passed base queue to pushReplicateChange, but only its state
changer was being used. This commit updates pushReplicateChange to only take in
the state changer.

Epic: none
Release note: none

wenyihu6 committed:
asim: pass queue names for logging

This commit passes the name of the queue making changes to pushReplicateChange
to improve logging.

Epic: none
Release note: none

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149904)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/kv/kvserver/asim/queue/replicate_queue.go` line 115 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/149904#-OV5TW5UEEkSIApi2RMY:-OV5TW5UEEkSIApi2RMZ:bix4n1x) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4f9a14ad7bd814237f0fc06e883a43da97d275c/pkg/kv/kvserver/asim/queue/replicate_queue.go#L115)):*
> ```Go
> 	rq.AddLogTag("tick", tick)
> 	ctx = rq.ResetAndAnnotateCtx(ctx)
> 	if rq.lastTick.After(rq.next) {
> ```

While making this commit, I was unsure about the purpose of `lastTick` here. We're advancing `next` to `lastTick` if `next` < `lastTick`. Since `lastTick` should be < `current tick`, `next` must also be < `current tick` and pass the check below on `!tick.Before(next)`. And this seems to be the only place we are checking for next tick. Am I missing something? 


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
> Reviewed 7 of 7 files at r1, 2 of 2 files at r2, 4 of 4 files at r3, 2 of 2 files at r4, 2 of 2 files at r5, 9 of 9 files at r6, 2 of 2 files at r7, 1 of 1 files at r8, 1 of 1 files at r9, 2 of 2 files at r10, 26 of 26 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149904)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg and @wenyihu6)

___
*[`pkg/kv/kvserver/asim/queue/replicate_queue.go` line 115 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/149904#-OV5TW5UEEkSIApi2RMY:-OV71mdm09YOI3j39SJV:bve1q1f) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4f9a14ad7bd814237f0fc06e883a43da97d275c/pkg/kv/kvserver/asim/queue/replicate_queue.go#L115)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

While making this commit, I was unsure about the purpose of `lastTick` here. We're advancing `next` to `lastTick` if `next` < `lastTick`. Since `lastTick` should be < `current tick`, `next` must also be < `current tick` and pass the check below on `!tick.Before(next)`. And this seems to be the only place we are checking for next tick. Am I missing something?
</blockquote></details>

It's fishy, you're right. I'm having trouble even understanding what exactly the semantics of the method are.

So we have a priority queue of ranges to take action on behalf of. Let's say we enter that for loop, we pull a rangeID off the queue, we push the replication change. Now (in your fixed version of the code, which seems to have been what was intended to happen in the old version), `rq.next` is the timestamp at which this (say) replication change takes effect: this will presumably be in the future, and the for loop terminates, we remember the last tick, and are done.

So when entering `Tick`, it's likely expected that `rq.next >= tick`. If the queue was empty recently and so `rq.next` hasn't been updated in a while, we could have `rq.next` trailing `tick`. What the point is of then making sure it's at least forwarded to `lastTick` beats me. The tests all passed with this removed (in all three queues).
I thought about just removing it and adding some commits, but figured this would possibly interfere with something else, so let's revisit that separately. For now, consider adding a TODO?

Btw, the only real use of `lastTick` that I can see is for the split queue, which needs to hand it to the load based splitter here: https://github.com/cockroachdb/cockroach/blob/5968b31ae888cc234e56144f4f741dfbc3417fa4/pkg/kv/kvserver/asim/queue/split_queue.go#L123

But this is more of a plumbing problem - the caller to `MaybeAdd` has the current tick (and has just ticked the queues, so `lastTick = tick`).
I wonder if it would make sense for the current tick to become a property of `State`, then it wouldn't need to be plumbed around, and semantically that also seems to make sense to me - the current time could be considered a property of the system.
___
*[`-- commits` line 112 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/149904#-OV7X6gS1seTKHiqUbGw:-OV7X6gS1seTKHiqUbGx:bbe8c6o):*
The mma datadriven tests are pretty messy. I'd like to follow this up with some separation between the two. Let's talk about this in our next meeting. It would be nice if the SMA tests would also run in "MMA" mode and could be compared directly in some quantitative way (and vice versa).
___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/example_rebalancing.txt` line 43 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/149904#-OV7VuAV8521PfFk5WiY:-OV7VuAW60DKoZWrPWbs:b-sm6hbg) ([raw file](https://github.com/cockroachdb/cockroach/blob/246d1caa08f43b8a652a0d39bd477fa074d559dc/pkg/kv/kvserver/asim/tests/testdata/non_rand/example_rebalancing.txt#L43)):*
> ```help
> # and stable. This should only be done for tests where the seed is given, so
> # the output is identical..
> plot stat=qps sample=2
> ```

I was running this example as of this commit and I'm seeing this throughout the entire test run:

```
I250714 10:09:13.164947 25 kv/kvserver/allocator/mmaprototype/allocator_state.go:868  [-] 795  did not register external changes: due to range does not exist in cluster state
```

This strikes me as surprising. Is it?
___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/example_rebalancing.txt` line 78 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/149904#-OV7O9R--tZLL4CfhFVJ:-OV7O9R--tZLL4CfhFVK:b6je5dj) ([raw file](https://github.com/cockroachdb/cockroach/blob/246d1caa08f43b8a652a0d39bd477fa074d559dc/pkg/kv/kvserver/asim/tests/testdata/non_rand/example_rebalancing.txt#L78)):*
> ```help
> failed assertion sample 2
>   balance stat=qps threshold=(<1.15) ticks=6
> 	max/mean=2.00 tick=0
> ```

Do we expect to hit this assertion? I guess we do, and the comment was already accurate while the datadriven output had lagged?

Can you help me understand why this particular commit got us the assertion failure? Are we now somehow accurately simulating the gossip delays?
___
*[`pkg/kv/kvserver/asim/queue/replicate_queue.go` line 180 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/149904#-OV7-DDS9odr_hc391Uh:-OV7-DDS9odr_hc391Ui:bnlu63d) ([raw file](https://github.com/cockroachdb/cockroach/blob/e120c953018a9bd7ae0ec8c6fcc117b784283d55/pkg/kv/kvserver/asim/queue/replicate_queue.go#L180)):*
> ```Go
> 			TransferTarget: state.StoreID(op.Target.StoreID),
> 			Author:         state.StoreID(op.Source.StoreID),
> 			// TODO(mma): Should this be add? I don't think so since it will assume
> ```

It makes sense to me to pass `false` here. So do we want to keep a TODO here? If anything, the TODO would've made sense before changing this (in which case the choice of `add=true` would've been questionable).
___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/example_skewed_cpu_even_ranges_mma` line 35 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/149904#-OV7Wm-iEjFh-hX8mIpm:-OV7Wm-iEjFh-hX8mIpn:b-tyykxf) ([raw file](https://github.com/cockroachdb/cockroach/blob/5968b31ae888cc234e56144f4f741dfbc3417fa4/pkg/kv/kvserver/asim/tests/testdata/non_rand/example_skewed_cpu_even_ranges_mma#L35)):*
> ```
> OK
> 
> print
> ```

Not for now, but these state printouts could be a lot more concise.
___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/example_tobi` line 0 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/149904#-OV7WuRz0n9sgku11kqY:-OV7WuRz0n9sgku11kqZ:b-1i4hf9) ([raw file](https://github.com/cockroachdb/cockroach/blob/5968b31ae888cc234e56144f4f741dfbc3417fa4/pkg/kv/kvserver/asim/tests/testdata/non_rand/example_tobi#L0)):*
Heh, `example_tobi` can probably go at some point :smile: 


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149904)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[`-- commits` line 112 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/149904#-OV7X6gS1seTKHiqUbGw:-OV91KS9933hBNwMUXlh:bjd2ma2):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

The mma datadriven tests are pretty messy. I'd like to follow this up with some separation between the two. Let's talk about this in our next meeting. It would be nice if the SMA tests would also run in "MMA" mode and could be compared directly in some quantitative way (and vice versa).
</blockquote></details>

Agreed, I dropped the last commit (which adds the datadriven tests and causes timeout) while we work on hardening the data driven tests tgtr. 
___
*[`pkg/kv/kvserver/asim/queue/replicate_queue.go` line 180 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/149904#-OV7-DDS9odr_hc391Uh:-OV91bXf3egpXv6YQrYS:bh85xp4) ([raw file](https://github.com/cockroachdb/cockroach/blob/e120c953018a9bd7ae0ec8c6fcc117b784283d55/pkg/kv/kvserver/asim/queue/replicate_queue.go#L180)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

It makes sense to me to pass `false` here. So do we want to keep a TODO here? If anything, the TODO would've made sense before changing this (in which case the choice of `add=true` would've been questionable).
</blockquote></details>

Agreed that this looks like a stale comment. Removed. 
___
*[`pkg/kv/kvserver/asim/queue/replicate_queue.go` line 115 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/149904#-OV5TW5UEEkSIApi2RMY:-OV91ZvTAMpbCtTiEfc7:betc06h) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4f9a14ad7bd814237f0fc06e883a43da97d275c/pkg/kv/kvserver/asim/queue/replicate_queue.go#L115)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

It's fishy, you're right. I'm having trouble even understanding what exactly the semantics of the method are.

So we have a priority queue of ranges to take action on behalf of. Let's say we enter that for loop, we pull a rangeID off the queue, we push the replication change. Now (in your fixed version of the code, which seems to have been what was intended to happen in the old version), `rq.next` is the timestamp at which this (say) replication change takes effect: this will presumably be in the future, and the for loop terminates, we remember the last tick, and are done.

So when entering `Tick`, it's likely expected that `rq.next >= tick`. If the queue was empty recently and so `rq.next` hasn't been updated in a while, we could have `rq.next` trailing `tick`. What the point is of then making sure it's at least forwarded to `lastTick` beats me. The tests all passed with this removed (in all three queues).
I thought about just removing it and adding some commits, but figured this would possibly interfere with something else, so let's revisit that separately. For now, consider adding a TODO?

Btw, the only real use of `lastTick` that I can see is for the split queue, which needs to hand it to the load based splitter here: https://github.com/cockroachdb/cockroach/blob/5968b31ae888cc234e56144f4f741dfbc3417fa4/pkg/kv/kvserver/asim/queue/split_queue.go#L123

But this is more of a plumbing problem - the caller to `MaybeAdd` has the current tick (and has just ticked the queues, so `lastTick = tick`).
I wonder if it would make sense for the current tick to become a property of `State`, then it wouldn't need to be plumbed around, and semantically that also seems to make sense to me - the current time could be considered a property of the system.
</blockquote></details>

Agreed, added TODO. 
___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/example_rebalancing.txt` line 43 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/149904#-OV7VuAV8521PfFk5WiY:-OV91pdV9DwCgWVvzKtQ:b-ydwaer) ([raw file](https://github.com/cockroachdb/cockroach/blob/246d1caa08f43b8a652a0d39bd477fa074d559dc/pkg/kv/kvserver/asim/tests/testdata/non_rand/example_rebalancing.txt#L43)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I was running this example as of this commit and I'm seeing this throughout the entire test run:

```
I250714 10:09:13.164947 25 kv/kvserver/allocator/mmaprototype/allocator_state.go:868  [-] 795  did not register external changes: due to range does not exist in cluster state
```

This strikes me as surprising. Is it?
</blockquote></details>

I think this was because MMA isn’t enabled in this test. Without it being enabled, MMA can’t register external changes because it never learns about the range via [`processStoreLeaseholderMsg`](https://github.com/cockroachdb/cockroach/blob/1841964459212e078605d1631de296dd1c03cb7e/pkg/kv/kvserver/asim/mmaintegration/mma_store_rebalancer.go#L156), and the this external change would get dropped https://github.com/cockroachdb/cockroach/blob/09ede0058ecea2dc331755de8462f7836ce87e6b/pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go#L1682. I expect most of our other tests without MMA enabled follow the same pattern. I’ll follow up with a PR to disable allocator sync when MMA is not enabled.
___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/example_rebalancing.txt` line 78 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/149904#-OV7O9R--tZLL4CfhFVJ:-OV92shw3irYr0UBVIZr:b-2rg20l) ([raw file](https://github.com/cockroachdb/cockroach/blob/246d1caa08f43b8a652a0d39bd477fa074d559dc/pkg/kv/kvserver/asim/tests/testdata/non_rand/example_rebalancing.txt#L78)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Do we expect to hit this assertion? I guess we do, and the comment was already accurate while the datadriven output had lagged?

Can you help me understand why this particular commit got us the assertion failure? Are we now somehow accurately simulating the gossip delays?
</blockquote></details>

This test passes if we increase `test_duration` to 30 minutes. The assertion failure was introduced in the commit `asim: correctly update queue.next in pushReplicateChange`. Makes sense on a high level since we now correctly wait for replication changes to complete before proceeding. So things appear to be slower now.
___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/example_skewed_cpu_even_ranges_mma` line 35 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/149904#-OV7Wm-iEjFh-hX8mIpm:-OV938z_0Fv88XmwgK-R:b2xiznq) ([raw file](https://github.com/cockroachdb/cockroach/blob/5968b31ae888cc234e56144f4f741dfbc3417fa4/pkg/kv/kvserver/asim/tests/testdata/non_rand/example_skewed_cpu_even_ranges_mma#L35)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Not for now, but these state printouts could be a lot more concise.
</blockquote></details>

Added a new item to our spreadsheet to track this. 
___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/example_tobi` line  at r11](https://reviewable.io/reviews/cockroachdb/cockroach/149904#-OV7WuRz0n9sgku11kqY:-OV93C7oEMNmmBKjtFi5:b6z4mo2) ([raw file](https://github.com/cockroachdb/cockroach/blob/5968b31ae888cc234e56144f4f741dfbc3417fa4/pkg/kv/kvserver/asim/tests/testdata/non_rand/example_tobi#L)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Heh, `example_tobi` can probably go at some point :smile:
</blockquote></details>

Removed the commit entirely - we can handle it in the follow-up.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
TFTR!

bors r=tbg

wenyihu6 commented:
TFTR!

bors r=tbg
